source/jormungandr/jormungandr/routing/valhalla.py

```diff
@@ -0,0 +1,190 @@
+
+            section.begin_date_time = journey.departure_date_time
+            section.end_date_time = journey.arrival_date_time
+
+            section.id = 'section_0'
+            section.duration = journey.duration
+            section.length = int(leg['summary']['length'] * coefficient)
+            section.origin.name = leg['maneuvers'][0]['street_names'][0]
+            section.origin.uri = '{};{}'.format(locations[0]['lon'], locations[0]['lat'])
+            section.origin.embedded_type = type_pb2.ADDRESS
+            section.origin.address.coord.lon = locations[0]['lon']
+            section.origin.address.coord.lat = locations[0]['lat']
+            section.origin.address.label = section.destination.name
+
+            section.destination.name = leg['maneuvers'][len(leg['maneuvers'])-2]['street_names'][0]
+            section.destination.uri = '{};{}'.format(locations[1]['lon'], locations[1]['lat'])
+            section.destination.embedded_type = type_pb2.ADDRESS
+            section.destination.address.coord.lon = locations[1]['lon']
+            section.destination.address.coord.lat = locations[1]['lat']
+            section.destination.address.label = section.destination.name
+
+            section.street_network.length = leg['summary']['length'] * coefficient
+            section.street_network.duration = leg['summary']['time']
+            section.street_network.mode = map_mode[mode]
+            for maneuver in leg['maneuvers']:
+                path_item = section.street_network.path_items.add()
+                if 'street_names' in maneuver and len(maneuver['street_names']) > 0:
+                    path_item.name = maneuver['street_names'][0]
+                path_item.length = maneuver['length'] * coefficient
+                path_item.duration = maneuver['time']
+
+            for sh in shape:
+                coord = section.street_network.coordinates.add()
+                coord.lon = sh[0]
+                coord.lat = sh[1]
+
+        return resp
+
+    def __format_url(self, mode, origin, destination):
+        map_mode = {
+            "walking": "pedestrian",
+            "car": "auto",
+            "bike": "bicycle"
+        }
+        if mode not in map_mode:
+            logging.getLogger(__name__).error('Valhalla, mode {} not implemented'.format(mode))
+            raise NotImplementedError()
+        url = {
+            "locations": [self.__format_coord(origin), self.__format_coord(destination)],
+            "costing": map_mode.get(mode),
+            "directions_options": self.args.get('service_args')
+        }
+        return '{}/route?json={}'.format(self.args.get('service_url'), json.dumps(url))
+
+    def __direct_path(self, mode, origin, destination, datetime, clockwise):
+        url = self.__format_url(mode, origin, destination)
+        r = self.__call_valhalla(url)
+        if not r:
+            raise TechnicalError('impossible to access valhalla service')
```

Comment:
It will be nice to at least subclass this exception at the abstract definition: every implementation must have the same interface.